tools/halide_image.h

```diff
@@ -1,278 +1,668 @@
-        std::vector<int> dimSizes(dimension_sizes(vals));
-        size_t dims = dimSizes.size();
-        assert(dims <= 4);
-        initialize
-            ( dims > 0 ? dimSizes[0] : 1
-            , dims > 1 ? dimSizes[1] : 0
-            , dims > 2 ? dimSizes[2] : 0
-            , dims > 3 ? dimSizes[3] : 0
-            , false);
-        int n = 1;
-        for (size_t i = 0; i < dims; ++i)
-            n *= dimSizes[i];
-        T const *ary = first_of_array(vals);
-        T *host = data();
-        for (int i = 0; i < n; ++i)
-            host[i] = ary[i];
-    }
-
-    /** Make sure you've called copy_to_host before you start
-     * accessing pixels directly. */
-    T &operator()(int x, int y = 0, int z = 0, int w = 0) {
-        T *ptr = (T *)contents->buf.host;
-        x -= contents->buf.min[0];
-        y -= contents->buf.min[1];
-        z -= contents->buf.min[2];
-        w -= contents->buf.min[3];
-        size_t s0 = contents->buf.stride[0];
-        size_t s1 = contents->buf.stride[1];
-        size_t s2 = contents->buf.stride[2];
-        size_t s3 = contents->buf.stride[3];
-        return ptr[s0 * x + s1 * y + s2 * z + s3 * w];
-    }
-
-    /** Make sure you've called copy_to_host before you start
-     * accessing pixels directly */
-    const T &operator()(int x, int y = 0, int z = 0, int w = 0) const {
-        const T *ptr = (const T *)contents->buf.host;
-        x -= contents->buf.min[0];
-        y -= contents->buf.min[1];
-        z -= contents->buf.min[2];
-        w -= contents->buf.min[3];
-        size_t s0 = contents->buf.stride[0];
-        size_t s1 = contents->buf.stride[1];
-        size_t s2 = contents->buf.stride[2];
-        size_t s3 = contents->buf.stride[3];
-        return ptr[s0 * x + s1 * y + s2 * z + s3 * w];
-    }
-
-    operator buffer_t *() const {
-        return &(contents->buf);
+    bool device_dirty() const {
+        return dev_dirty;
     }
 
-    int width() const {
-        return dimensions() > 0 ? contents->buf.extent[0] : 1;
+    // In the future there will also be a host_dirty() method, but
+    // currently it conflicts with a buffer_t field.
+
+    void set_device_dirty(bool v = true) {
+        dev_dirty = v;
     }
 
-    int height() const {
-        return dimensions() > 1 ? contents->buf.extent[1] : 1;
+    void copy_to_host() {
+        if (device_dirty()) {
+            halide_copy_to_host(NULL, this);
+        }
     }
 
-    int channels() const {
-        return dimensions() > 2 ? contents->buf.extent[2] : 1;
+    void copy_to_device(const struct halide_device_interface_t *device_interface) {
+        if (host_dirty()) {
+            halide_copy_to_device(NULL, this, device_interface);
+        }
     }
 
-    int dimensions() const {
-        for (int i = 0; i < 4; i++) {
-            if (contents->buf.extent[i] == 0) {
-                return i;
-            }
+    void device_free() {
+        halide_device_free(nullptr, this);
```

Comment:
I did consider adding a device_free to the custom deleter inside the
shared_ptr, but at that point the buffer is already gone, and the buffer
may refer to a device allocation even when it does not own a host
allocation.... suggestions for the right ownership semantics are welcome

On Fri, Jul 22, 2016 at 4:10 PM, Dillon Sharlet <notifications@github.com>
wrote:

> In tools/halide_image.h
> <https://github.com/halide/Halide/pull/1402#discussion_r71957962>:
>
> >      }
> >
> > -    int dimensions() const {
> > -        for (int i = 0; i < 4; i++) {
> > -            if (contents->buf.extent[i] == 0) {
> > -                return i;
> > -            }
> > +    void device_free() {
> > +        halide_device_free(nullptr, this);
>
> It seems difficult to automatically call this without doing something
> tricky (a struct that is wrapped in std::shared_ptr, like alloc)... but it
> would be nice to find a way to do this.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/halide/Halide/pull/1402/files/db7c75d1d239d3e130905aa6363cf5e9f0e8a150#r71957962>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAfdRoqo1Yl5zaL1QQs27WIhcO18wdk_ks5qYU3fgaJpZM4JTMHs>
> .
>